src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/RewriteRuleExecutor.java

```diff
@@ -0,0 +1,68 @@
+/*
+ *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.rewriter;
+
+import com.alibaba.druid.sql.ast.expr.SQLQueryExpr;
+
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Query RewriteRuleExecutor which will execute the {@link RewriteRule} with registered order.
+ */
+public class RewriteRuleExecutor<T extends SQLQueryExpr> {
```

Comment:
I think this class should not be generic at all, only the internal collection should be generic, because that way you can have rules for different types of expressions instead of one fixed type.